RAQUASI88/src/WIN32/retroachievements.cpp

```diff
@@ -180,18 +210,107 @@ void RA_InitMemory()
 #endif
 }
 
-void RA_OnGameClose()
+int RA_PrepareLoadNewRom(const char *file_name, int file_type)
 {
-    RA_ClearMemoryBanks();
-    RA_SetGameTitle("");
-    loaded_title[0] = '\0';
-    RA_OnLoadNewRom(NULL, 0);
+    FILE *f = fopen(file_name, "rb");
+
+    char basename[_MAX_FNAME];
+    _splitpath(file_name, NULL, NULL, basename, NULL);
+    strcpy(loading_file.name, basename);
+
+    fseek(f, 0, SEEK_END);
+    const unsigned long file_size = (unsigned long)ftell(f);
+    loading_file.data_len = file_size;
+
+    BYTE * const file_data = (BYTE *)malloc(file_size * sizeof(BYTE));
+    loading_file.data = file_data;
+    fseek(f, 0, SEEK_SET);
+    fread(file_data, sizeof(BYTE), file_size, f);
+
+    fflush(f);
+    fclose(f);
+
+    loading_file.title_id = RA_IdentifyRom(file_data, file_size);
+    loading_file.file_type = file_type;
+
+    if (loaded_title != NULL && loaded_title->data_len > 0)
+    {
+        if (loaded_title->title_id != loading_file.title_id || loaded_title->file_type != loading_file.file_type)
+        {
+            if (!RA_WarnDisableHardcore("load a new title without ejecting all images and resetting the emulator"))
+            {
+                free_file_info(&loading_file);
+                return FALSE; /* 読み込みを中止する */
+            }
+        }
+
+        RA_ConfirmLoadNewRom(false);
+    }
+
+    return TRUE;
+}
+
+void RA_CommitLoadNewRom()
+{
+    switch (loading_file.file_type)
+    {
+    case FTYPE_DISK:
+        free_file_info(&loaded_disk);
+        loaded_disk = loading_file;
+        loaded_title = &loaded_disk;
+        break;
+    case FTYPE_TAPE_LOAD:
+        free_file_info(&loaded_tape);
+        loaded_tape = loading_file;
+        loaded_title = &loaded_tape;
+        break;
+    default:
+        break;
+    }
+
+    RA_UpdateAppTitle(loading_file.name);
+
+    /* 実績システムのイメージデータを初期化する */
+    RA_ActivateGame(loading_file.title_id);
```

Comment:
I will think about it, but actually it might be better to reactivate, since memory gets reinitialized between disk loads. The main goal is to block title-swapping in Hardcore mode.

Reply:
Handling it via a preprocessor flag set above `RA_PrepareLoadNewRom()`.